### PR TITLE
feat: use events to decouple services

### LINKED ISSUE #232: Refactor: decouple ParticipantContextService using events
# Feature Request

The `ParticipantContextService` currently interacts with collaborators (`DidDocumentService`, `Vault` etc) via composition. this may lead to a tight coupling, so we'd rather have them interact via (synchronous transactional) events.

## Which Areas Would Be Affected?

_e.g., DPF, CI, build, transfer, etc._

## Why Is the Feature Desired?

decoupling, flexibility

## Solution Proposal

The `ParticipantContextService` emits an event whenever a `ParticipantContext` object is modified to allow collaborator services to interact.

### PR DESCRIPTION
**Pull request by paullatzelsperger**
## What this PR changes/adds

This PR introduces events, which are emitted by services, to decrease service coupling.

Specifically these services emit events:
- `participantContextService`
- `KeyPairService`
- `VerifiableCredentialService`

## Why it does that

Decrease service coupling

## Further notes

_List other areas of code that have changed but are not necessarily linked to the main feature. This could be method
signature changes, package declarations, bugs that were encountered and were fixed inline, etc._

## Linked Issue(s)

Closes #232

_Please be sure to take a look at the [contributing guidelines](https://github.com/eclipse-edc/.github/blob/main/CONTRIBUTING.md#submit-a-pull-request) and our [etiquette for pull requests](https://github.com/eclipse-edc/.github/blob/main/contributing/pr_etiquette.md)._

### CODE REVIEW

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java [hunk 1]
```diff
@@ -181,9 +194,77 @@ public ServiceResult<Void> removeService(String did, String serviceId) {
...
+    public <E extends Event> void on(EventEnvelope<E> eventEnvelope) {
+        var payload = eventEnvelope.getPayload();
+        if (payload instanceof ParticipantContextCreated event) {
+            created(event);
+        } else if (payload instanceof ParticipantContextUpdated event) {
+            updated(event);
+        } else if (payload instanceof ParticipantContextDeleted event) {
+            deleted(event);
+        } else {
+            monitor.warning("KeyPairServiceImpl Received event with unexpected payload type: %s".formatted(payload.getClass()));
```
Comment (jimmarino) [inline core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:213]:
```suggestion
            monitor.warning("KeyPairServiceImpl Received an event with unexpected payload type: %s".formatted(payload.getClass()));
```

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java [hunk 2]
```diff
@@ -181,9 +194,77 @@ public ServiceResult<Void> removeService(String did, String serviceId) {
...
+                    case ACTIVATED -> publish(resource.getDid());
+                    case DEACTIVATED -> unpublish(resource.getDid());
+                    default -> ServiceResult.success();
+                })
+                .filter(AbstractResult::failed)
+                .map(AbstractResult::getFailureDetail)
+                .collect(Collectors.joining(", "));
+
+        if (!errors.isEmpty()) {
+            monitor.warning("Updating DID documents subsequent to updating a ParticipantContext failed: %s".formatted(errors));
```
Comment (jimmarino) [inline core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:232]:
```suggestion
            monitor.warning("Updating DID documents after updating a ParticipantContext failed: %s".formatted(errors));
```

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java [hunk 3]
```diff
@@ -181,9 +194,77 @@ public ServiceResult<Void> removeService(String did, String serviceId) {
...
+        //unpublish and delete all DIDs associated with that participant
+        var errors = findByParticipantId(participantId)
+                .stream()
+                .map(didResource -> unpublish(didResource.getDid())
+                        .compose(u -> deleteById(didResource.getDid())))
+                .map(AbstractResult::getFailureDetail)
+                .collect(Collectors.joining(", "));
+
+        if (!errors.isEmpty()) {
+            monitor.warning("Unpublishing/deleting DID documents subsequent to deleting a ParticipantContext failed: %s".formatted(errors));
```
Comment (jimmarino) [inline core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:247]:
```suggestion
            monitor.warning("Unpublishing/deleting DID documents after deleting a ParticipantContext failed: %s".formatted(errors));
```

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java [hunk 4]
```diff
@@ -181,9 +194,77 @@ public ServiceResult<Void> removeService(String did, String serviceId) {
...
+                .id(manifest.getDid())
+                .service(manifest.getServiceEndpoints().stream().toList())
+                // updating and adding a verification method happens as a result of the KeyPairAddedEvent
+                // .verificationMethod(List.of(VerificationMethod.Builder.newInstance()
+                // .publicKeyJwk(publicKey.toJSONObject())
+                // .build()))
+                .build();
+        store(doc, manifest.getParticipantId())
+                .compose(u -> manifest.isActive() ? publish(doc.getId()) : success())
+                .onFailure(f -> monitor.warning("Creating a DID document subsequent to a ParticipantContext creation failed: %s".formatted(f.getFailureDetail())));
```
Comment (jimmarino) [inline core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:268]:
```suggestion
                .onFailure(f -> monitor.warning("Creating a DID document after creating a ParticipantContext creation failed: %s".formatted(f.getFailureDetail())));
```

File: spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/diddocument/DidDocumentEvent.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+package org.eclipse.edc.identityhub.spi.events.diddocument;
+
+import org.eclipse.edc.spi.event.Event;
+
+import java.util.Objects;
+
+/**
+ * Base class for all events related to activites and state changes of DID documents
```
Comment (jimmarino) [inline spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/diddocument/DidDocumentEvent.java:22]:
```suggestion
 * Base class for all events related to DID document activites and state changes.
```

File: spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/participant/ParticipantContextDeleted.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Event that signals that a {@link org.eclipse.edc.identityhub.spi.model.participant.ParticipantContext} was deleted
+ */
+public class ParticipantContextDeleted extends ParticipantContextEvent {
+    @Override
+    public String name() {
+        return "participantcontext.deleted";
+    }
+
+    @JsonPOJOBuilder(withPrefix = "")
```
Comment (jimmarino) [inline spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/participant/ParticipantContextDeleted.java:29]: For all classes with @JsonPOJOBuilder, we should implement full serialization params or remove this param (this is on several event classes)

File: spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/participant/ParticipantContextListener.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+     * can now happen.
+     *
+     * @param newContext The newly created (already persisted) participant context
+     * @param manifest   The original manifest based on which the context was created
+     */
+    default void created(ParticipantContext newContext, ParticipantManifest manifest) {
+
+    }
+    
+    default void updated(ParticipantContext updatedContext) {
```
Comment (jimmarino) [inline spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/participant/ParticipantContextListener.java:38]: Doc

File: spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/participant/ParticipantContextListener.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    }
+
```
Comment (jimmarino) [inline spi/identity-hub-spi/src/main/java/org/eclipse/edc/identityhub/spi/events/participant/ParticipantContextListener.java:41]: Doc

**Review by jimmarino — COMMENTED** [review]
Some wording comments. I also noticed there are some incomplete Jackson serialization  annotations - these should either be removed or full annotations and tests should be added.

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java [hunk 9]
```diff
@@ -181,9 +194,77 @@ public ServiceResult<Void> removeService(String did, String serviceId) {
...
+
+    private void created(ParticipantContextCreated event) {
+        var manifest = event.getManifest();
+        var doc = DidDocument.Builder.newInstance()
+                .id(manifest.getDid())
+                .service(manifest.getServiceEndpoints().stream().toList())
+                // updating and adding a verification method happens as a result of the KeyPairAddedEvent
+                // .verificationMethod(List.of(VerificationMethod.Builder.newInstance()
+                // .publicKeyJwk(publicKey.toJSONObject())
+                // .build()))
```
Comment (wolf4ood) [inline core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:264]: to remove ?

**Review by wolf4ood — APPROVED** [review]

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java (hunk 9 above)
Comment (paullatzelsperger) [reply core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:264]: i noticed, that the code add verification methods was actually missing 🤦. Will add that and then remove the commented code here.

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java [hunk 10]
```diff
@@ -209,11 +219,61 @@ public <E extends Event> void on(EventEnvelope<E> eventEnvelope) {
...
+                .map(AbstractResult::getFailureDetail)
+                .collect(Collectors.joining(","));
+
+        if (!errors.isEmpty()) {
+            monitor.warning("Updating DID documents after adding a KeyPair failed: %s".formatted(errors));
+        }
+    }
+
+    private void keypairAdded(KeyPairAdded event) {
+        var diddocs = findByParticipantId(event.getParticipantId());
```
Comment (jimmarino) [inline core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:248]: diddocs --> didDocs

File: core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java [hunk 11]
```diff
@@ -209,11 +219,61 @@ public <E extends Event> void on(EventEnvelope<E> eventEnvelope) {
...
+        var publicKey = keyParserRegistry.parse(serialized);
+
+        if (publicKey.failed()) {
+            monitor.warning("Error adding KeyPair '%s' to DID Document of participant '%s': %s".formatted(event.getKeyId(), event.getParticipantId(), publicKey.getFailureDetail()));
+            return;
+        }
+
+        var jwk = CryptoConverter.createJwk(new KeyPair((PublicKey) publicKey.getContent(), null));
+
+
```
Comment (jimmarino) [inline core/identity-hub-did/src/main/java/org/eclipse/edc/identityhub/did/DidDocumentServiceImpl.java:259]: spacing

**Review by jimmarino — APPROVED** [review]

**Pull request merged**